File: src/protocol.rs
use serde::{Deserialize, Serialize};

use crate::game::Snapshot;

#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientMsg {
    Join {
        player_id: String,
        name: String,
        spectator: bool,
    },
    SetTicket {
        ticket: String,
    },
    Vote {
        card: String,
    },
    Reveal,
    PickEstimate,
    SetEstimate {
        card: String,
    },
    ConfirmRound,
    Revote,
    SetSpectator {
        spectator: bool,
    },
    SetDeck {
        deck: String,
    },
}

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMsg {
    Welcome { player_id: String, state: Snapshot },
    State { state: Snapshot },
    Error { message: String },
}

#[derive(Debug, Clone, Deserialize)]
pub struct CreateBoardRequest {
    pub name: Option<String>,
    pub deck: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BoardMeta {
    pub id: String,
    pub name: String,
    pub deck: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct DeckInfo {
    pub id: String,
    pub label: String,
    pub preview: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HistoryResponse {
    pub id: String,
    pub name: String,
    pub rounds: Vec<crate::game::CompletedRound>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ErrorBody {
    pub error: String,
}

File: shared/protocol.ts
export type Phase = "voting" | "revealed" | "choosing";

export type PlayerView = {
  id: string;
  name: string;
  spectator: boolean;
  connected: boolean;
  has_voted: boolean;
  vote: string | null;
};

export type VoteRecord = {
  name: string;
  card: string | null;
};

export type CompletedRound = {
  id: string;
  ticket: string;
  agreed: string;
  votes: VoteRecord[];
  completed_at: string;
};

export type Snapshot = {
  board_id: string;
  board_name: string;
  deck_id: string;
  cards: string[];
  ticket: string;
  phase: Phase;
  proposed_estimate: string | null;
  players: PlayerView[];
  you: PlayerView;
  completed: CompletedRound[];
  average: number | null;
};

export type ClientMsg =
  | { type: "join"; player_id: string; name: string; spectator: boolean }
  | { type: "set_ticket"; ticket: string }
  | { type: "vote"; card: string }
  | { type: "reveal" }
  | { type: "pick_estimate" }
  | { type: "set_estimate"; card: string }
  | { type: "confirm_round" }
  | { type: "revote" }
  | { type: "set_spectator"; spectator: boolean }
  | { type: "set_deck"; deck: string };

export type ServerMsg =
  | { type: "welcome"; player_id: string; state: Snapshot }
  | { type: "state"; state: Snapshot }
  | { type: "error"; message: string };

export type HistoryResponse = {
  id: string;
  name: string;
  rounds: CompletedRound[];
};

export type BoardMeta = {
  id: string;
  name: string;
  deck: string;
};

export type DeckInfo = {
  id: string;
  label: string;
  preview: string;
};

export type CreateBoardRequest = {
  name?: string;
  deck: string;
};
